### ISSUE TITLE
Regression for TypeVar inference in nested function call.

### ISSUE THREAD

**Issue by huntzhan** [microsoft/pyright#4463]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

We have a automatic procedure to validate and upgrade pyright to the latest version. And we discovered the a type inference regression from version `288` to `289`(Related PR  https://github.com/vkit-x/vkit/pull/94. Just FYI, no need to click). Looks like this issue is related to https://github.com/microsoft/pyright/issues/4457, but I'm not sure.

**To Reproduce**
Steps to reproduce the behavior.

```python
from typing import TypeVar, Optional


class Foo:
    pass


class Bar:
    pass


_T = TypeVar('_T', Foo, Bar)


def func_shared(some_obj: Optional[_T]):
    return 1


def func0(some_obj: Optional[_T]):
    assert some_obj
    return func_shared(some_obj)


def func1(some_obj: Optional[_T]):
    assert some_obj
    return func_shared(some_obj)


def debug():
    func0(Foo())
    func1(Bar())
```

**Expected behavior**

Aforementioned example should pass type checking.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.

But instead, pyright raise the following error:

```bash
❯ pyright tests/debug_pyright.py
Loading configuration file at /Users/huntzhan/Project/personal/vkit-dev/vkit/pyrightconfig.json
stubPath /Users/huntzhan/Project/personal/vkit-dev/vkit/.github/typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.289
/Users/huntzhan/Project/personal/vkit-dev/vkit/tests/debug_pyright.py
  /Users/huntzhan/Project/personal/vkit-dev/vkit/tests/debug_pyright.py:22:24 - error: Argument of type "Foo* | Bar*" cannot be assigned to parameter "some_obj" of type "_T@func_shared | None" in function "func_shared"
    Type "Foo* | Bar*" cannot be assigned to type "Bar* | None"
      Type "Foo*" cannot be assigned to type "Bar* | None"
        "Foo*" is incompatible with "Bar*"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
  /Users/huntzhan/Project/personal/vkit-dev/vkit/tests/debug_pyright.py:27:24 - error: Argument of type "Foo* | Bar*" cannot be assigned to parameter "some_obj" of type "_T@func_shared | None" in function "func_shared"
    Type "Foo* | Bar*" cannot be assigned to type "Bar* | None"
      Type "Foo*" cannot be assigned to type "Bar* | None"
        "Foo*" is incompatible with "Bar*"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
Completed in 0.574sec
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

Command line. The latest version `289`.

**Additional context**

Aforementioned context should be enough.

**Comment by erictraut** [comment]
This was due to a change targeted at fixing https://github.com/microsoft/pyright/issues/4439. I'll need to see if there's a way to address both patterns.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.290, which I just published. It will also be included in a future release of pylance.
